### PR TITLE
Laboratory work #1, Anna Podpryatova - 22FPL2

### PR DESCRIPTION
**Pull request by AnPruch**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -0,0 +1 @@
+
```
Comment (artyomtugaryov) [inline main.py:1]: remove the file

File: seminars/practice_2_string.py [hunk 2]
```diff
@@ -53,13 +53,14 @@ def multiply_string(input_string: str, how_many: int) -> str:
...
     # student realisation goes here
+    return how_many * input_string
 
 
 # Function calls with expected result:
-# multiply_string('Hi', 2) → 'HiHi'
-# multiply_string('Hi', 3) → 'HiHiHi'
-# multiply_string('Hi', 1) → 'Hi'
-# multiply_string('Hi', 0) → ''
+multiply_string('Hi', 2) # → 'HiHi'
```
Comment (artyomtugaryov) [inline seminars/practice_2_string.py:60]: discard

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Perfect start! Please, fix CI (all red markers)

File: seminars/practice_2_string.py (hunk 2 above)
Comment (AnPruch) [reply seminars/practice_2_string.py]: Done

File: main.py (hunk 1 above)
Comment (AnPruch) [reply main.py]: Done

**Comment by AnPruch** [comment]
> Perfect start! Please, fix CI (all red markers)

Could you please help me to understand some of the checks?

**Import style checks** tells me that my imports are in the wrong position or missing. Should I import every function from main.py in a specific order that I don't know, even if I don't reach for some functions in start.py explicitly or what should I do?

File: seminars/practice_2_string.py [hunk 3]
```diff
@@ -53,13 +53,14 @@ def multiply_string(input_string: str, how_many: int) -> str:
...
     # student realisation goes here
+    # return how_many * input_string
 
 
 # Function calls with expected result:
-# multiply_string('Hi', 2) → 'HiHi'
-# multiply_string('Hi', 3) → 'HiHiHi'
-# multiply_string('Hi', 1) → 'Hi'
-# multiply_string('Hi', 0) → ''
+# multiply_string('Hi', 2) # → 'HiHi'
```
Comment (artyomtugaryov) [inline seminars/practice_2_string.py:60]: Revert this changes

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_1_classify_by_unigrams/start.py [hunk 4]
```diff
@@ -1,20 +1,48 @@
-"""
-Language detection starter
-"""
+"""Language detection starter"""
+from lab_1_classify_by_unigrams.main import create_language_profile, detect_language_advanced, \
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/start.py:2]: 1. try to add empty line after the doc string
2. Revert doc string to the previous version (multiline)
3. try to use `(` and `)` instead of `\`

Comment (AnPruch) [reply lab_1_classify_by_unigrams/start.py]: I did. Still it doesn't work

Comment (AnPruch) [reply lab_1_classify_by_unigrams/start.py]: I did. Still it doesn't work

File: lab_1_classify_by_unigrams/main.py [hunk 5]
```diff
@@ -11,6 +12,12 @@ def tokenize(text: str) -> list[str] | None:
     :param text: a text
     :return: a list of lower-cased tokens without punctuation
     """
+    if not isinstance(text, str):
+        return None
+
+    tokens = [character for character in text.lower() if character.isalpha()]
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:18]: perfect, lets remove it without creating the variable

File: lab_1_classify_by_unigrams/main.py [hunk 6]
```diff
@@ -49,6 +85,31 @@ def compare_profiles(
...
     """
+    if not isinstance(unknown_profile, dict) or not isinstance(profile_to_compare, dict) \
+            or ('name' or 'freq') not in unknown_profile \
+            or ('name' or 'freq') not in profile_to_compare:
+        return None
+
+    unknown_freq = unknown_profile['freq']
+    compare_freq = profile_to_compare['freq']
+    mutual_characters = [[], []]
+    if isinstance(unknown_freq, dict) and \
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:96]: reverse the condition and return None as soon as posible

File: lab_1_classify_by_unigrams/main.py [hunk 7]
```diff
@@ -71,6 +150,14 @@ def load_profile(path_to_file: str) -> dict | None:
...
     :return: a dictionary with at least two keys – name, freq
     """
+    if not isinstance(path_to_file, str):
+        return None
+
+    with open(path_to_file, 'r', encoding='utf-8') as file:
+        profile = json.load(file)
+        if isinstance(profile, dict):
+            return profile
+        return None
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:160]: returns - outside of `with`

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Very good

File: lab_1_classify_by_unigrams/main.py (hunk 7 above)
Comment (AnPruch) [reply lab_1_classify_by_unigrams/main.py]: Corrected

File: lab_1_classify_by_unigrams/main.py (hunk 6 above)
Comment (AnPruch) [reply lab_1_classify_by_unigrams/main.py]: Reversed

File: lab_1_classify_by_unigrams/main.py (hunk 5 above)
Comment (AnPruch) [reply lab_1_classify_by_unigrams/main.py]: Did it. Hopefully, in the way you meant it

File: lab_1_classify_by_unigrams/main.py [hunk 8]
```diff
@@ -71,6 +149,15 @@ def load_profile(path_to_file: str) -> dict | None:
     :param path_to_file: a path to the language profile
     :return: a dictionary with at least two keys – name, freq
     """
+    if not isinstance(path_to_file, str):
+        return None
+
+    with open(path_to_file, 'r', encoding='utf-8') as file:
+        profile = json.load(file)
+
+    if isinstance(profile, dict):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:158]: the same

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Very good

File: lab_1_classify_by_unigrams/main.py (hunk 8 above)
Comment (AnPruch) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py [hunk 9]
```diff
@@ -19,6 +24,16 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
...
     :return: a dictionary with frequencies
     """
+    if not isinstance(tokens, list) or not all(isinstance(char, str) for char in tokens):
+        return None
+
+    text_freq = {}
+    for character in tokens:
+        if character not in text_freq:
+            text_freq[character] = 0.0
+        text_freq[character] += 1/len(tokens)
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:34]: not enought spaces

File: lab_1_classify_by_unigrams/main.py [hunk 10]
```diff
@@ -37,6 +61,16 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
...
     """
+    if not isinstance(predicted, list) or not isinstance(actual, list) \
+            or len(predicted) != len(actual):
+        return None
+
+    mse = 0.0
+    for index, pred_value in enumerate(predicted):
+        mse += (pred_value - actual[index])**2
+
+    mse /= len(predicted)
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:72]: do not create the function

File: lab_1_classify_by_unigrams/main.py [hunk 11]
```diff
@@ -63,6 +122,24 @@ def detect_language(
...
+        return None
+
+    mse_1 = compare_profiles(unknown_profile, profile_1)
+    mse_2 = compare_profiles(unknown_profile, profile_2)
+
+    if not mse_1 and not mse_2:
+        return None
+    if mse_2 and (not mse_1 or mse_1 > mse_2) and \
+            isinstance(profile_2['name'], str):
+        return profile_2['name']
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:136]: profile_2['name'] / profile_1['name'] to variables, please

File: lab_1_classify_by_unigrams/main.py [hunk 12]
```diff
@@ -80,6 +166,22 @@ def preprocess_profile(profile: dict) -> dict[str, str | dict] | None:
...
+            or 'n_words' not in profile or 'freq' not in profile:
+        return None
+
+    perfect_profile = {'name': profile['name'],
+                       'freq': {}}
+
+    for sequence in profile['freq']:
+        if sequence.lower() in perfect_profile['freq']:
+            perfect_profile['freq'][sequence.lower()] += \
+                profile['freq'][sequence] / profile['n_words'][0]
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:179]: `profile['freq'][sequence]` / `profile['n_words'][0]` to variables

**Review by artyomtugaryov — APPROVED** [review]
Nice!

File: lab_1_classify_by_unigrams/main.py (hunk 10 above)
Comment (AnPruch) [reply lab_1_classify_by_unigrams/main.py:72]: What do you mean?

File: lab_1_classify_by_unigrams/main.py (hunk 12 above)
Comment (AnPruch) [reply lab_1_classify_by_unigrams/main.py:179]: Done

File: lab_1_classify_by_unigrams/main.py (hunk 9 above)
Comment (AnPruch) [reply lab_1_classify_by_unigrams/main.py]: Done

File: lab_1_classify_by_unigrams/main.py (hunk 10 above)
Comment (AnPruch) [reply lab_1_classify_by_unigrams/main.py]: I removed the line 72 and returned without reassigning mse
